### PR TITLE
[FEAT] Public Delta Lake writer

### PR DESCRIPTION
**Pull request by kevinzwang**
Follow-up on #2304 with additional cleanup, parameters, and testing of the Delta Lake writing functionality, now set to be a public API with this PR!

This PR also renames `read_delta_lake` to `read_deltalake`, providing a deprecation warning for the old name

Todo before merging:
- [x] Test on S3/AWS Glue
- [x] Add GIthub issues for future work
  - #2332

### CODE REVIEW

**Comment by kevinzwang** [comment]
Thank you to @siddharth-gulia for the original work on this!

**Comment by kevinzwang** [comment]
Oh one question I had is if it should be kept as `write_delta`? The read function is called `read_delta_lake` and they should probably match. I would guess that this was probably named `write_delta` to match Polars

File: daft/dataframe/dataframe.py [hunk 1]
```diff
@@ -549,62 +549,109 @@ def write_iceberg(self, table: "pyiceberg.table.Table", mode: str = "append") ->
         # This is due to the fact that the logical plan of the write_iceberg returns datafiles but we want to return the above data
         return with_operations
 
+    @DataframePublicAPI
     def write_delta(
         self,
-        path: str,
-        mode: str = "append",
+        table: Union[str, pathlib.Path, "deltalake.DeltaTable"],
```
Comment (samster25) [inline daft/dataframe/dataframe.py:555]: This should also be able to take in a `daft.io.catalog.DataCatalogTable` which lets users use their unity catalog

File: daft/dataframe/dataframe.py [hunk 2]
```diff
@@ -549,62 +549,109 @@ def write_iceberg(self, table: "pyiceberg.table.Table", mode: str = "append") ->
         # This is due to the fact that the logical plan of the write_iceberg returns datafiles but we want to return the above data
         return with_operations
 
+    @DataframePublicAPI
     def write_delta(
```
Comment (samster25) [inline daft/dataframe/dataframe.py:553]: I think this should be consistent with `read_delta_lake` so lets rename to `write_delta_lake`?

File: daft/execution/execution_step.py [hunk 3]
```diff
@@ -412,7 +412,6 @@ def _handle_file_write(self, input: MicroPartition) -> MicroPartition:
 class WriteDeltaLake(SingleOutputInstruction):
     base_path: str
     large_dtypes: bool
-    current_version: int
```
Comment (samster25) [inline daft/execution/execution_step.py:415]: I would leave this in to enable overwrites at a per partition level like specified here https://docs.databricks.com/en/delta/selective-overwrite.html#dynamic-partition-overwrites

**Review by samster25 — APPROVED** [review]
Discussed offline, but main thing is to keep the io related changes in `daft/io` and rely on dynamic imports when io is invoked

**Pull request merged**